packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -5140,26 +5130,29 @@ class StatefulElement extends ComponentElement {
     final StatefulWidget oldWidget = state._widget!;
     state._widget = widget as StatefulWidget;
     final Object? debugCheckForReturnedFuture = state.didUpdateWidget(oldWidget) as dynamic;
-    assert(() {
-      if (debugCheckForReturnedFuture is Future) {
-        throw FlutterError.fromParts(<DiagnosticsNode>[
-          ErrorSummary('${state.runtimeType}.didUpdateWidget() returned a Future.'),
-          ErrorDescription( 'State.didUpdateWidget() must be a void method without an `async` keyword.'),
-          ErrorHint(
-            'Rather than awaiting on asynchronous work directly inside of didUpdateWidget, '
-            'call a separate method to do this work without awaiting it.',
-          ),
-        ]);
-      }
-      return true;
-    }());
+    assert(_debugCheckReturnValueNotFuture(debugCheckForReturnedFuture, 'didUpdateWidget'));
     rebuild(force: true);
   }
 
+  bool _debugCheckReturnValueNotFuture(Object? value, String name) {
+    if (value is Future) {
```

Comment:
You should wrap this all in an assert so that it doesn't do anything in non-debug builds.  I know this is only called within asserts, but just in case it ever gets called outside an assert...

Alternatively, you could set a sentinel value here in an assert block, then check that the sentinel value is set -- to keep people from accidentally calling this in non-debug builds.

Reply:
> So shall I modify all of them using your suggestion as well?

No it's not worth the churn.  A quick grep shows that some of them use this style (e.g. [here](https://github.com/flutter/flutter/blob/eaf625448899555cff6736fe030c1064abd79ff3/packages/flutter/lib/src/widgets/framework.dart#L5843)) and some don't (e.g. [here](https://github.com/flutter/flutter/blob/eaf625448899555cff6736fe030c1064abd79ff3/packages/flutter/lib/src/material/slider_theme.dart#L3475)).  But let's use the style in this case.

Reply:
Thanks for the review @tvolkert! 

@fzyzcjy do you plan to follow up on this suggestion?

Reply:
Ah I am quite busy doing research now, but I will come back when having a vacation!